File: backend-go/internal/modules/system/service/online_user_service.go
package service

import (
	"context"
	"encoding/json"
	"sort"
	"strings"

	"drip-admin/backend-go/internal/common"
	"github.com/gin-gonic/gin"
)

func (s *Server) OnlineUsers(c *gin.Context) {
	page, err := common.ParsePage(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	rows, err := s.onlineSessionRows(c.Request.Context())
	if err != nil {
		common.HandleError(c, err)
		return
	}
	rows = filterOnlineRows(rows, "username", c.Query("username"))
	rows = filterOnlineRows(rows, "ip", c.Query("ip"))
	rows = filterOnlineRows(rows, "deviceType", c.Query("deviceType"))
	start := (page.Page - 1) * page.PageSize
	if start > len(rows) {
		start = len(rows)
	}
	end := start + page.PageSize
	if end > len(rows) {
		end = len(rows)
	}
	currentToken := c.GetHeader(s.cfg.Token.Name)
	out := make([]OnlineUserVo, 0, end-start)
	for _, row := range rows[start:end] {
		out = append(out, row.toVo(currentToken))
	}
	common.Success(c, common.PageResult[OnlineUserVo]{List: out, Total: common.Int64String(len(rows)), Page: page.Page, PageSize: page.PageSize})
}

func (s *Server) OnlineUser(c *gin.Context) {
	tokenID := c.Param("tokenId")
	session, err := s.sessionByToken(c.Request.Context(), tokenID)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	if session.TokenID == "" {
		common.HandleError(c, common.NewBusinessError(404000, "online session not found"))
		return
	}
	common.Success(c, session.toVo(c.GetHeader(s.cfg.Token.Name)))
}

func (s *Server) KickoutOnlineUser(c *gin.Context) {
	tokenID := c.Param("tokenId")
	if tokenID == c.GetHeader(s.cfg.Token.Name) {
		common.HandleError(c, common.NewBusinessError(400000, "operation failed"))
		return
	}
	successOrError(c, nil, s.removeSession(c.Request.Context(), tokenID))
}

func (session SessionData) toVo(currentToken string) OnlineUserVo {
	return OnlineUserVo{TokenID: session.TokenID, UserID: session.UserID, Username: session.Username, RealName: session.RealName, DeviceType: session.DeviceType, IP: session.IP, UserAgent: session.UserAgent, LoginAt: session.LoginAt, LastActiveAt: session.LastActiveAt, ExpireAt: session.ExpireAt, Current: session.TokenID == currentToken}
}

func (s *Server) onlineSessionRows(ctx context.Context) ([]SessionData, error) {
	rows := []SessionData{}
	iter := s.redis.Scan(ctx, 0, sessionKeyPrefix+"*", 100).Iterator()
	for iter.Next(ctx) {
		value, err := s.redis.Get(ctx, iter.Val()).Result()
		if err != nil {
			continue
		}
		var row SessionData
		if err := json.Unmarshal([]byte(value), &row); err == nil {
			rows = append(rows, row)
		}
	}
	if err := iter.Err(); err != nil {
		return nil, common.NewBusinessError(500000, "failed to read online session")
	}
	sort.Slice(rows, func(i, j int) bool { return rows[i].LastActiveAt > rows[j].LastActiveAt })
	return rows, nil
}

func filterOnlineRows(rows []SessionData, field string, value string) []SessionData {
	if strings.TrimSpace(value) == "" {
		return rows
	}
	needle := strings.ToLower(strings.TrimSpace(value))
	out := rows[:0]
	for _, row := range rows {
		var current string
		switch field {
		case "username":
			current = row.Username
		case "ip":
			current = row.IP
		case "deviceType":
			current = row.DeviceType
		}
		if strings.Contains(strings.ToLower(current), needle) {
			out = append(out, row)
		}
	}
	return out
}

File: backend-python/app/modules/system/service/online_user_service.py
import json

from redis.asyncio import Redis

from app.common.errors import bad_request, not_found
from app.common.pagination import PageQuery, PageResult
from app.modules.system.service.auth_service import SESSION_PREFIX, TOKEN_PREFIX


class OnlineUserService:
    def __init__(self, redis: Redis) -> None:
        self.redis = redis

    async def list_users(
        self,
        page: PageQuery,
        current_token: str,
        username: str | None,
        ip: str | None,
        device_type: str | None,
    ) -> PageResult[dict]:
        sessions = await self._sessions()
        sessions = [
            item
            for item in sessions
            if _contains(item.get("username"), username)
            and _contains(item.get("ip"), ip)
            and _contains(item.get("deviceType"), device_type)
        ]
        sessions.sort(key=lambda item: item.get("lastActiveAt", ""), reverse=True)
        total = len(sessions)
        start = (page.page - 1) * page.page_size
        rows = sessions[start : start + page.page_size]
        return PageResult(
            list=[self._vo(item, current_token) for item in rows],
            total=str(total),
            page=page.page,
            pageSize=page.page_size,
        )

    async def detail(self, token_id: str, current_token: str) -> dict:
        session = await self._session_by_token(token_id)
        if session is None:
            raise not_found("online session not found")
        return self._vo(session, current_token)

    async def kickout(self, token_id: str, current_token: str) -> None:
        if token_id == current_token:
            raise bad_request("operation failed")
        session = await self._session_by_token(token_id)
        if session is None:
            raise not_found("online session not found")
        await self.redis.delete(
            f"{TOKEN_PREFIX}{token_id}",
            f"{SESSION_PREFIX}{session['userId']}:{session['deviceType']}",
        )

    async def _sessions(self) -> list[dict]:
        rows: list[dict] = []
        async for key in self.redis.scan_iter(match=f"{SESSION_PREFIX}*"):
            payload = await self.redis.get(key)
            if not payload:
                continue
            try:
                rows.append(json.loads(payload))
            except json.JSONDecodeError:
                await self.redis.delete(key)
        return rows

    async def _session_by_token(self, token_id: str) -> dict | None:
        payload = await self.redis.get(f"{TOKEN_PREFIX}{token_id}")
        if not payload:
            return None
        try:
            return json.loads(payload)
        except json.JSONDecodeError:
            await self.redis.delete(f"{TOKEN_PREFIX}{token_id}")
            return None

    @staticmethod
    def _vo(session: dict, current_token: str) -> dict:
        return {
            "tokenId": session["tokenId"],
            "userId": str(session["userId"]),
            "username": session["username"],
            "realName": session.get("realName", ""),
            "deviceType": session.get("deviceType", ""),
            "ip": session.get("ip", ""),
            "userAgent": session.get("userAgent", ""),
            "loginAt": session.get("loginAt", ""),
            "lastActiveAt": session.get("lastActiveAt", ""),
            "expireAt": session.get("expireAt", ""),
            "current": session["tokenId"] == current_token,
        }


def _contains(current: str | None, expected: str | None) -> bool:
    if not expected or not expected.strip():
        return True
    return expected.strip().lower() in (current or "").lower()
